### PR TITLE
[rtsan][NFC] Documentation of suppression flag

### PR DESCRIPTION
**Pull request by cjappl**

### CODE REVIEW

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [inline clang/docs/RealtimeSanitizer.rst:220]: I appreciate that we might want something really concise and snappy here, but if that's not a massive motivation, I think we could quantify these costs a bit, rather than just saying low, medium and high. Maybe something like "Very low: single bool check" could work?

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [inline clang/docs/RealtimeSanitizer.rst:221]: I wonder whether the word "Suppresses" is a bit loaded in this context and could lead the reader down the wrong path a little. I worry the user might put this in the same mental box as the suppressions file. Perhaps that's not a problem.  

How about something more like "- All violations are ignored for the lifetime of the disabler"?

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [inline clang/docs/RealtimeSanitizer.rst:225]: What about "- Medium: single regex match"?

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [inline clang/docs/RealtimeSanitizer.rst:230]: "High: unwind stack, symbolize and regex match each frame"? Possibly a bit long - not sure how you feel about this.

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [inline clang/docs/RealtimeSanitizer.rst:237]: "all sanitizer error reports are ~~suppressed~~ *ignored*", maybe?

File: clang/docs/RealtimeSanitizer.rst
Comment (cjappl) [reply clang/docs/RealtimeSanitizer.rst:220]: Yes, I'm amenable to this, but I started down this path then aborted.

My worry was providing too much low level unnecessary detail to the user, when a high level overview would suffice.

Given the info about what the actual cost is (vs "very low") how would I, the user, behave differently? Is additional (possibly complex to understand) knowledge about what is happening under the hood helpful? (honest question, I originally decided "no" but have definitely been turning this over in my head.

File: clang/docs/RealtimeSanitizer.rst
Comment (cjappl) [reply clang/docs/RealtimeSanitizer.rst:221]: Great thought. I'm not sure where I come down on it, but I'm also amenable to this change.

I guess the question is (similar to the above) "Is there a difference between suppressing and ignoring an issue? As a user do I care?"

I could see one case for calling everything suppressing (which is what I inadvertently did in this PR), or make the differentiation between suppressing (what a suppression file does) and ignoring (what the Disabler does).

To the end user, they do the same thing, to us, we behave differently and go through different mechanisms to achieve that same result.

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [reply clang/docs/RealtimeSanitizer.rst:220]: Totally understood, and well phrased question - is what is happening under the hood helpful? In general it depends on the context, of course. For something like choosing between `std::map` vs `std::unordered_map` for performance reasons, it's critical to know what's happening under the hood. For choosing what json library to use in an application that isn't performance sensitive, it doesn't matter - I just care about the API being easy to use. So where does rtsan fall on that spectrum? I think probably somewhere in the middle, but I suspect leaning more towards the "I don't care" side. So maybe we just leave this as it is, and add more detail if people request it? I don't feel strongly either way on this one, just wanted to raise the question in case it rung some alarm bells.

Comment (cjappl) [reply clang/docs/RealtimeSanitizer.rst:220]: Yeah, I think it's good that you raised it, especially because I had the exact same back and forth. I'll leave this conversation as unresolved and maybe fmayer has an opinion.

I lean towards keeping it simple like this, but a very slight lean.

Comment (fmayer) [reply clang/docs/RealtimeSanitizer.rst:220]: I would also not add too many implementation details into the public documentation. It's unlikely to be useful, not unlikely to be wrong at some point in the future when the implementation changes.

Comment (davidtrevelyan) [reply clang/docs/RealtimeSanitizer.rst:220]: Great - agreed. I'm happy to leave as it is and for this comment to be resolved 👍

**Review by fmayer — APPROVED** [review]

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [reply clang/docs/RealtimeSanitizer.rst]: Nice thought. I was marginally in favour of "ignored", but your point here makes me lean further towards "ignored". I feel like the user's use cases mean:

- using ScopedDisabler is the user saying "turn off realtimesanitizer"
- using suppressions means "keep using rtsan - hide these errors, but keep note of them"

Definitely support splitting the terminology here, but not ultra-strongly if anyone feels otherwise

Comment (cjappl) [reply clang/docs/RealtimeSanitizer.rst]: Switched it. Please have a read through of the latest commit and make sure everything is good - switched some wording around minorly to make the grammar work.

**Comment by cjappl** [comment]
(test failure seems unrelated, but I will ensure it's green before I submit after I get approval, going to rebase now)

File: clang/docs/RealtimeSanitizer.rst
Comment (davidtrevelyan) [inline clang/docs/RealtimeSanitizer.rst:237]: Minor nits: not sure the word "potential" adds value here (it raises questions), and I think "for that thread" and "invoked functions" are a given as we're already talking about a variable scope. I'd propose a simplification as follows:

_All RTSan errors originating within the ``ScopedDisabler`` instance variable scope are ignored._

I don't feel strongly though, if you feel this is less clear - no worries.

**Review by davidtrevelyan — APPROVED** [review]

File: clang/docs/RealtimeSanitizer.rst
Comment (cjappl) [reply clang/docs/RealtimeSanitizer.rst]: Agreed this sentence was getting a bit overwrought. Check out the latest, reworded yours a bit.

File: clang/docs/RealtimeSanitizer.rst
Comment (cjappl) [reply clang/docs/RealtimeSanitizer.rst]: Did one last revision on this like we discussed on our call. I'll submit this later if tests pass

**Pull request merged**
